python/cudf/cudf/core/column/categorical.py

```diff
@@ -191,108 +208,137 @@ def _set_categories(self, new_categories, **kwargs):
         df = df.sort_values(by="order").reset_index(True)
 
         ordered = kwargs.get("ordered", self.ordered)
-        kwargs = df["new_codes"]._column._replace_defaults()
-        kwargs.update(categories=new_cats, ordered=ordered, name=None)
+        new_codes = df["new_codes"]._column
+        new_dtype = CategoricalDtype(
+            data_dtype=new_codes.dtype, categories=new_cats, ordered=ordered,
+        )
 
         if kwargs.get("inplace", False):
-            self._parent._categories = new_cats
-            self._parent._mask = kwargs["mask"]
-            self._parent._data = kwargs["data"]
-            self._parent._ordered = kwargs["ordered"]
-            self._parent._null_count = kwargs["null_count"]
-            self._parent._dtype = CategoricalDtype(ordered=ordered)
+            self._parent.categories = new_cats
+            self._parent.mask = new_codes.mask
+            self._parent.data = new_codes.data
+            self._parent.ordered = ordered
+            self._parent.dtype = new_dtype
             return None
 
-        return self._parent.replace(**kwargs)
+        return column.build_column(
+            data=new_codes.data, dtype=new_dtype, mask=new_codes.mask
+        )
 
 
-class CategoricalColumn(column.TypedColumnBase):
+class CategoricalColumn(column.ColumnBase):
     """Implements operations for Columns of Categorical type
     """
 
-    def __init__(self, **kwargs):
+    def __init__(self, data, dtype, mask=None, name=None):
```

Comment:
`data` should be `None`, while `children` holds two columns for the codes and categories - similar to how string columns are

Reply:
`dtype` can just be `category`

Reply:
Done